File: services/discovery-service/internal/events/events.go
package events

import (
	"context"
	"encoding/json"
	"time"

	"github.com/google/uuid"
	"github.com/rs/zerolog/log"
	"github.com/segmentio/kafka-go"
)

type CloudEvent struct {
	ID          string      `json:"id"`
	Source      string      `json:"source"`
	SpecVersion string      `json:"specversion"`
	Type        string      `json:"type"`
	Time        string      `json:"time"`
	Data        interface{} `json:"data"`
}

type KafkaProducer struct {
	writer *kafka.Writer
}

func NewKafkaProducer(brokers []string) *KafkaProducer {
	return &KafkaProducer{
		writer: &kafka.Writer{
			Addr:     kafka.TCP(brokers...),
			Topic:    "discovery-events",
			Balancer: &kafka.LeastBytes{},
		},
	}
}

func (p *KafkaProducer) Publish(ctx context.Context, eventType string, data interface{}) {
	event := CloudEvent{
		ID:          uuid.New().String(),
		Source:      "discovery-service",
		SpecVersion: "1.0",
		Type:        eventType,
		Time:        time.Now().UTC().Format(time.RFC3339),
		Data:        data,
	}
	body, err := json.Marshal(event)
	if err != nil {
		log.Error().Err(err).Str("type", eventType).Msg("failed to marshal event")
		return
	}
	if err := p.writer.WriteMessages(ctx, kafka.Message{Value: body}); err != nil {
		log.Error().Err(err).Str("type", eventType).Msg("failed to publish event")
		return
	}
	log.Info().Str("type", eventType).Msg("event published")
}

func (p *KafkaProducer) Close() {
	if p.writer != nil {
		p.writer.Close()
	}
}

type NoopProducer struct{}

func NewNoopProducer() *NoopProducer {
	return &NoopProducer{}
}

func (p *NoopProducer) Publish(ctx context.Context, eventType string, data interface{}) {
	log.Info().Str("type", eventType).Msg("noop event")
}

func (p *NoopProducer) Close() {}

File: services/media-service/src/events.rs
use async_trait::async_trait;
use serde_json::Value;

#[async_trait]
pub trait EventProducer: Send + Sync {
    async fn publish(&self, event_type: &str, data: Value);
    async fn close(&self);
}

pub struct NoopProducer;

#[async_trait]
impl EventProducer for NoopProducer {
    async fn publish(&self, event_type: &str, data: Value) {
        tracing::info!(event_type, data = %data, "noop event");
    }

    async fn close(&self) {}
}

#[cfg(feature = "kafka")]
mod kafka_impl {
    use super::*;
    use rdkafka::producer::{FutureProducer, FutureRecord};
    use rdkafka::ClientConfig;
    use std::time::Duration;

    pub struct KafkaProducer {
        producer: FutureProducer,
        topic: String,
    }

    impl KafkaProducer {
        pub fn new(brokers: &str) -> Self {
            let producer: FutureProducer = ClientConfig::new()
                .set("bootstrap.servers", brokers)
                .set("message.timeout.ms", "5000")
                .create()
                .expect("failed to create kafka producer");
            Self {
                producer,
                topic: "media-events".into(),
            }
        }
    }

    #[async_trait]
    impl EventProducer for KafkaProducer {
        async fn publish(&self, event_type: &str, data: Value) {
            let event = serde_json::json!({
                "id": uuid::Uuid::new_v4().to_string(),
                "source": "media-service",
                "specversion": "1.0",
                "type": event_type,
                "time": chrono::Utc::now().to_rfc3339(),
                "data": data,
            });
            let payload = serde_json::to_string(&event).unwrap_or_default();
            let record = FutureRecord::to(&self.topic).payload(&payload).key(&event_type);
            match self.producer.send(record, Duration::from_secs(5)).await {
                Ok(_) => tracing::info!(event_type, "event published"),
                Err((e, _)) => tracing::error!(error = %e, "failed to publish event"),
            }
        }

        async fn close(&self) {}
    }
}

#[cfg(feature = "kafka")]
pub use kafka_impl::KafkaProducer;
